examples/getting_started/web_search.ipynb

```diff
@@ -0,0 +1,255 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "!pip install \"mirascope[openai]\" requests beautifulsoup4 pydantic"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "import os\n",
+        "\n",
+        "import requests\n",
+        "from bs4 import BeautifulSoup\n",
+        "from openai.types.chat import ChatCompletionMessageParam\n",
+        "from pydantic import BaseModel\n",
+        "\n",
+        "from mirascope.core import openai, prompt_template\n",
+        "\n",
+        "# Set your OpenAI API key\n",
+        "os.environ[\"OPENAI_API_KEY\"] = \"your-api-key-here\""
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 3. Implementing the Tools <a name=\"tools\"></a>\n",
+        "\n",
+        "Now, let's implement our tools: `web_search` and `parse_webpage`. These functions will allow our agent to search the web and extract content from webpages."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "def web_search(query: str) -> str:\n",
+        "    \"\"\"Perform a web search and return the results.\n",
+        "\n",
+        "    This is a simplified implementation. In a real-world scenario,\n",
+        "    you would integrate with a proper search engine API.\n",
+        "    \"\"\"\n",
+        "    # Placeholder implementation\n",
+        "    return f\"Search results for: {query}\"\n",
+        "\n",
+        "\n",
+        "def parse_webpage(url: str) -> str:\n",
+        "    \"\"\"Parse the content of a webpage.\"\"\"\n",
+        "    try:\n",
+        "        response = requests.get(url)\n",
+        "        soup = BeautifulSoup(response.content, \"html.parser\")\n",
+        "        return \"\\n\".join([p.text for p in soup.find_all(\"p\")])\n",
+        "    except Exception as e:\n",
+        "        return f\"Error parsing webpage: {str(e)}\""
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 4. Building the Web Search Agent <a name=\"building\"></a>\n",
+        "\n",
+        "Now that we have our tools ready, let's build the WebSearchAgent class. This class will use the Mirascope framework to create an AI agent that can interact with users, perform web searches, and provide informative responses."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "MAX_RESULTS = 10\n",
+        "\n",
+        "\n",
+        "class WebSearchAgent(BaseModel):\n",
+        "    history: list[ChatCompletionMessageParam] = []\n",
```

Comment:
nit: see recent PR for updating this to use `BaseMessageParam | openai.OpenAIMessageParam`